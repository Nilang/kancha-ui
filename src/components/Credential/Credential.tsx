import * as React from 'react'
import Container from '../Container/Container'
import Text, { TextTypes } from '../Text/Text'
import Avatar from '../Avatar/Avatar'
import Card from '../Card/Card'
import Icon from '../Icon/Icon'
import * as Kancha from '../../types'
import { withTheme } from '../../theming'
import S from 'string'

export interface CredentialProps {
  onPress?: () => void
  issuer: Kancha.Identity
  subject: Kancha.Identity
  exp: number
  fields: any
  testID?: string
  shadow?: number
  background?: Kancha.BrandPropOptions
  theme: any
}

const Credential: React.FC<CredentialProps> = ({
  shadow,
  onPress,
  issuer,
  background,
  fields,
  subject,
  testID,
  theme,
}) => {
  const subProfileSource = subject.profileImage ? { source: { uri: subject.profileImage } } : {}
  const issProfileSource = issuer.profileImage ? { source: { uri: issuer.profileImage } } : {}
  const credentialFields = Object.keys(fields)

  return (
    <Card onPress={onPress} testID={testID} shadow={shadow || 0} background={background}>
      <Container flexDirection={'row'} alignItems={'flex-start'}>
        <Container>
          <Avatar
            border
            {...subProfileSource}
            address={subject.did}
            type={'circle'}
            gravatarType={'retro'}
            size={40}
          />
          <Container viewStyle={{ position: 'absolute', top: 0, left: 15 }}>
            <Avatar
              border
              {...issProfileSource}
              address={issuer.did}
              type={'circle'}
              gravatarType={'retro'}
              size={40}
            />
          </Container>
        </Container>
        <Container marginLeft={28}>
          <Container>
            <Text bold type={TextTypes.ActivityTitle}>
              {issuer.shortId}
            </Text>
            <Container marginTop={3} flexDirection={'row'} alignItems={'flex-start'}>
              <Container marginRight={5}>
                <Icon
                  icon={{ name: 'ios-play', iconFamily: 'Ionicons' }}
                  size={15}
                  color={theme.colors.primary.text}
                />
              </Container>
              <Text type={TextTypes.ActivityTitle} textStyle={{ fontSize: 15 }}>
                {subject.shortId}
              </Text>
            </Container>
          </Container>
        </Container>
      </Container>
      <Container marginTop>
        {credentialFields.map((key: any, i: number) => {
          const fieldValueImage = !fields[key].isObj
            ? fields[key].value.endsWith('.jpg') || fields[key].value.endsWith('.png')
            : false
          return (
            i < 2 && (
              <Container marginBottom={5} flex={1} alignItems={'flex-start'} key={i}>
                <Container flex={1}>
                  <Text textStyle={{ fontSize: 12, textTransform: 'uppercase' }} type={TextTypes.SubTitle}>
                    {S(fields[key].type).humanize().s}
                  </Text>
                </Container>
                <Container justifyContent={'flex-start'} flex={2}>
                  {fieldValueImage ? (
                    <Container paddingTop={5}>
                      <Avatar
                        source={{ uri: fields[key].value }}
                        address={issuer.did}
                        type={'rounded'}
                        gravatarType={'retro'}
                        size={25}
                      />
                    </Container>
                  ) : (
                    <Text type={TextTypes.ActivityTitle}>
                      {fields[key].isObj ? 'Type not supported yet' : fields[key].value}
                    </Text>
                  )}
                </Container>
              </Container>
            )
          )
        })}
        {credentialFields.length > 2 && (
          <Container>
            <Text type={TextTypes.SubTitle}>...</Text>
          </Container>
        )}
      </Container>
    </Card>
  )
}

export default withTheme(Credential)
