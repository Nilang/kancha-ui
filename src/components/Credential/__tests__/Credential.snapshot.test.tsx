import * as React from 'react'
import { render, fireEvent } from 'react-native-testing-library'
import Credential from '../Credential'

describe('Component(snapshot): Credential', () => {
  it('should render with default props', () => {
    const tree = render(
      <Credential title={'Test Title'} issuer={'Test Issuer'} logo={{ uri: '' }} />,
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('should fire on press event', () => {
    const onPress = jest.fn()
    const { getByTestId } = render(
      <Credential
        title={'Test Title'}
        issuer={'Test Issuer'}
        logo={{ uri: '' }}
        onPress={onPress}
        testID={'CREDENTIAL_BTN'}
      />,
    )

    fireEvent.press(getByTestId('CREDENTIAL_BTN'))
    expect(onPress).toHaveBeenCalled()
  })
})
