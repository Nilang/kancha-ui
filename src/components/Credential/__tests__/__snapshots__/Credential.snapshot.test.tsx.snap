// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component(snapshot): Credential should render with default props 1`] = `
<View
  accessible={true}
  focusable={false}
  isTVSelectable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "elevation": 1,
      "shadowColor": "#000000",
      "shadowOpacity": 0.2,
      "shadowRadius": 5,
    }
  }
>
  <View
    pointerEvents="auto"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": "#ffffff",
        "borderRadius": 5,
        "elevation": 10,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": 15,
        "marginLeft": undefined,
        "marginRight": undefined,
        "marginTop": undefined,
        "padding": 15,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "shadowColor": "#ffffff",
        "shadowOpacity": 0.2,
        "shadowRadius": 5,
        "width": undefined,
      }
    }
  >
    <View
      pointerEvents="auto"
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": undefined,
          "borderRadius": undefined,
          "flex": undefined,
          "flexDirection": "row",
          "height": undefined,
          "justifyContent": undefined,
          "margin": undefined,
          "marginBottom": undefined,
          "marginLeft": undefined,
          "marginRight": undefined,
          "marginTop": undefined,
          "padding": undefined,
          "paddingBottom": undefined,
          "paddingHorizontal": undefined,
          "paddingLeft": undefined,
          "paddingRight": undefined,
          "paddingTop": undefined,
          "width": undefined,
        }
      }
    >
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": "#EAEAEA",
            "borderRadius": undefined,
            "flex": undefined,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": undefined,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "backgroundColor": "#0078ff",
              "borderRadius": 5,
              "height": 40,
              "width": 40,
            }
          }
        />
      </View>
      <View
        pointerEvents="auto"
        style={
          Object {
            "alignItems": undefined,
            "backgroundColor": undefined,
            "borderRadius": undefined,
            "flex": 1,
            "flexDirection": undefined,
            "height": undefined,
            "justifyContent": undefined,
            "margin": undefined,
            "marginBottom": undefined,
            "marginLeft": 15,
            "marginRight": undefined,
            "marginTop": undefined,
            "padding": undefined,
            "paddingBottom": undefined,
            "paddingHorizontal": undefined,
            "paddingLeft": undefined,
            "paddingRight": undefined,
            "paddingTop": undefined,
            "width": undefined,
          }
        }
      >
        <View
          pointerEvents="auto"
          style={
            Object {
              "alignItems": undefined,
              "backgroundColor": undefined,
              "borderRadius": undefined,
              "flex": undefined,
              "flexDirection": undefined,
              "height": undefined,
              "justifyContent": undefined,
              "margin": undefined,
              "marginBottom": undefined,
              "marginLeft": undefined,
              "marginRight": undefined,
              "marginTop": undefined,
              "padding": undefined,
              "paddingBottom": 3,
              "paddingHorizontal": undefined,
              "paddingLeft": undefined,
              "paddingRight": undefined,
              "paddingTop": undefined,
              "width": undefined,
            }
          }
        >
          <Text
            style={
              Object {
                "color": "#333333",
                "flexWrap": "wrap",
                "fontSize": 18,
                "fontWeight": "bold",
              }
            }
          >
            Test Title
          </Text>
        </View>
        <Text
          style={
            Object {
              "color": "#BBBBBB",
              "flexWrap": "wrap",
              "fontSize": 14,
            }
          }
        >
          Test Issuer
        </Text>
      </View>
    </View>
  </View>
</View>
`;
